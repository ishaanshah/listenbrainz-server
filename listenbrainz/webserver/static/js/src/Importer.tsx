import * as ReactDOM from "react-dom";
import * as React from "react";
import LastFMImporter from "./LastFMImporter";

import ImporterModal from "./ImporterModal";

export type ImporterProps = {
  user: {
    id?: string;
    name: string;
    auth_token: string;
  };
  profileUrl?: string;
  apiUrl?: string;
  lastfmApiUrl: string;
  lastfmApiKey: string;
};

export type ImporterState = {
  show: boolean;
  canClose: boolean;
  lastfmUsername: string;
  msg: string;
};

export default class Importer extends React.Component<
  ImporterProps,
  ImporterState
> {
  importer: any;

  constructor(props: ImporterProps) {
    super(props);

    this.state = {
      show: false,
      canClose: true,
      lastfmUsername: "",
      msg: "",
    };
  }

  handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ lastfmUsername: event.target.value });
  };

  handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    const { lastfmUsername } = this.state;
    this.toggleModal();
    event.preventDefault();
    this.importer = new LastFMImporter(lastfmUsername, this.props);
    setInterval(this.updateMessage, 100);
    setInterval(this.setClose, 100);
    this.importer.startImport();
  };

  toggleModal = () => {
    this.setState((prevState) => {
      return { show: !prevState.show };
    });
  };

  setClose = () => {
    this.setState({ canClose: this.importer.canClose });
  };

  updateMessage = () => {
    this.setState({ msg: this.importer.msg });
  };

  render() {
    const { show, canClose, lastfmUsername, msg } = this.state;

    return (
      <>
        <p>Most users will want to import from Last.fm directly.</p>
        <h3>Direct import from Last.fm</h3>
        <p>
          The Last.fm importer manually steps through your listen history and
          imports the listens one page at a time. Should it fail for whatever
          reason, it is safe to restart the import process. Running the import
          process multiple times <strong>does not</strong> create duplicates in
          your ListenBrainz listen history.
        </p>
        <p>
          In order for this to work, you must disable the &#34;Hide recent
          listening information&#34; setting in your Last.fm{" "}
          <a href="https://www.last.fm/settings/privacy">Privacy Settings</a>.
        </p>
        <p>
          Clicking the &quot;Import now!&qout; button will import your profile
          now without the need to open lastfm.
          <br />
          You need to keep this page open for the tool to work, it might take a
          while to complete. Though, you can continue doing your work. :)
        </p>
        <p>We need to know your Last.fm username:</p>
        <div className="well">
          <div className="Importer">
            <form onSubmit={this.handleSubmit}>
              <input
                type="text"
                onChange={this.handleChange}
                value={lastfmUsername}
                placeholder="Last.fm Username"
                size={30}
              />
              <input
                type="submit"
                value="Import Now!"
                disabled={!lastfmUsername}
              />
            </form>
            {show && (
              <ImporterModal onClose={this.toggleModal} disable={!canClose}>
                <img
                  src="/static/img/listenbrainz-logo.svg"
                  height="75"
                  className="img-responsive"
                  alt=""
                />
                <br />
                <br />
                <div>{msg}</div>
                <br />
              </ImporterModal>
            )}
          </div>
        </div>

        <h3> Import from Spotify </h3>
        <p>
          ListenBrainz can automatically import songs from Spotify as you listen
          to them.
        </p>
        <p>
          Importing the same listens from two different sources such as Last.FM
          and Spotify may cause the creation of duplicates in your listen
          history. If you opt into our automatic Spotify import, please do not
          use the Last.FM import or submit listens from other ListenBrainz
          clients. This is a temporary limitation while we find better ways to
          deduplicate listens.
        </p>
        <p>
          <a href={`{window.location.origin}/profile/connect-spotify`}>
            Connect your Spotify account to ListenBrainz.
          </a>
        </p>
      </>
    );
  }
}

document.addEventListener("DOMContentLoaded", () => {
  const domContainer = document.querySelector("#react-container");
  const propsElement = document.getElementById("react-props");
  let reactProps;
  try {
    reactProps = JSON.parse(propsElement!.innerHTML);
  } catch (err) {
    // Show error to the user and ask to reload page
  }
  const {
    user,
    profile_url,
    api_url,
    lastfm_api_url,
    lastfm_api_key,
  } = reactProps;
  ReactDOM.render(
    <Importer
      user={user}
      profileUrl={profile_url}
      apiUrl={api_url}
      lastfmApiKey={lastfm_api_key}
      lastfmApiUrl={lastfm_api_url}
    />,
    domContainer
  );
});
