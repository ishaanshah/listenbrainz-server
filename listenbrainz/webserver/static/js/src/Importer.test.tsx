import * as React from "react";
import { mount, shallow } from "enzyme";

import Importer from "./Importer";

const props = {
  user: {
    id: "id",
    name: "dummyUser",
    auth_token: "foobar",
  },
  profileUrl: "http://profile",
  apiUrl: "apiUrl",
  lastfmApiUrl: "http://ws.audioscrobbler.com/2.0/",
  lastfmApiKey: "foobar",
};

describe("Importer page", () => {
  it("renders", () => {
    const wrapper = mount(<Importer {...props} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("modal renders when button clicked", () => {
    const wrapper = shallow(<Importer {...props} />);
    // Simulate submiting the form
    wrapper.find("form").simulate("submit", {
      preventDefault: () => null,
    });

    // Test if the show property has been set to true
    expect(wrapper.exists("ImporterModal")).toBe(true);
  });

  it("submit button is disabled when input is empty", () => {
    const wrapper = shallow(<Importer {...props} />);
    // Make sure that the input is empty
    wrapper.setState({ lastfmUsername: "" });

    // Test if button is disabled
    expect(wrapper.find('input[type="submit"]').props().disabled).toBe(true);
  });
});
