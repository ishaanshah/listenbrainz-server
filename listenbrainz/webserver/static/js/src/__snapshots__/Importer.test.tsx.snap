// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Importer page renders 1`] = `"<p>Most users will want to import from Last.fm directly.</p><h3>Direct import from Last.fm</h3><p>The Last.fm importer manually steps through your listen history and imports the listens one page at a time. Should it fail for whatever reason, it is safe to restart the import process. Running the import process multiple times <strong>does not</strong> create duplicates in your ListenBrainz listen history.</p><p>In order for this to work, you must disable the \\"Hide recent listening information\\" setting in your Last.fm <a href=\\"https://www.last.fm/settings/privacy\\">Privacy Settings</a>.</p><p>Clicking the \\"Import now!&amp;qout; button will import your profile now without the need to open lastfm.<br>You need to keep this page open for the tool to work, it might take a while to complete. Though, you can continue doing your work. :)</p><p>We need to know your Last.fm username:</p><div class=\\"well\\"><div class=\\"Importer\\"><form><input type=\\"text\\" placeholder=\\"Last.fm Username\\" size=\\"30\\" value=\\"\\"><input type=\\"submit\\" disabled=\\"\\" value=\\"Import Now!\\"></form></div></div><h3> Import from Spotify </h3><p>ListenBrainz can automatically import songs from Spotify as you listen to them.</p><p>Importing the same listens from two different sources such as Last.FM and Spotify may cause the creation of duplicates in your listen history. If you opt into our automatic Spotify import, please do not use the Last.FM import or submit listens from other ListenBrainz clients. This is a temporary limitation while we find better ways to deduplicate listens.</p><p><a href=\\"{window.location.origin}/profile/connect-spotify\\">Connect your Spotify account to ListenBrainz.</a></p>"`;
